File: LeetCode_Google/Next_Permutation/Next_Permutation.cpp
// Next_Permutation.cpp
//
// Compile locally with:
// clang++ Next_Permutation.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    void nextPermutation(vector<int> &nums)
    {
        int n = nums.size();
        int left = n - 2;
        int right = n - 1;
        int tmp = 0;
        while (left >= 0 && nums[left] >= nums[left + 1])
        {
            left--;
        }
        if (left < 0)
        {
            left = 0;
            while (left < right)
            {
                tmp = nums[right];
                nums[right] = nums[left];
                nums[left] = tmp;
                left++;
                right--;
            }
            return;
        }

        right = left + 1;
        for (int index = right; index < n; index++)
        {
            if (nums[left] < nums[index])
            {
                right = index;
            }
        }
        tmp = nums[right];
        nums[right] = nums[left];
        nums[left] = tmp;

        left++;
        right = n - 1;
        while (left < right)
        {
            tmp = nums[right];
            nums[right] = nums[left];
            nums[left] = tmp;
            left++;
            right--;
        }
    }
};

int main()
{
    cout << "Running Next_Permutation.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}

File: LeetCode_Google/Next_Permutation/next__permutation.py
from typing import Optional, List
import heapq
from collections import defaultdict


class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """

        def reverse(a, b):
            while a < b:
                nums[a], nums[b] = nums[b], nums[a]
                a += 1
                b -= 1

        # find pivot
        left = len(nums) - 2

        while left >= 0 and nums[left] >= nums[left + 1]:
            left -= 1

        # if no pivot exists, return original
        if left < 0:
            reverse(0, len(nums) - 1)
            return nums

        # find largest index l with l>k such that nums[k] < nums[l]
        right = left + 1
        for tmp in range(right, len(nums)):
            if nums[left] < nums[tmp]:
                right = tmp

        nums[left], nums[right] = nums[right], nums[left]

        # reverse left + 1 to the end

        r1 = left + 1
        r2 = len(nums) - 1
        reverse(r1, r2)


if __name__ == "__main__":
    sol = Solution()

    nums = [4, 3, 2, 5, 4, 3, 1]
    print(sol.nextPermutation(nums))
    print("Running Solution...")
